pkg/apis/discovery/validation/validation.go

```diff
@@ -44,22 +53,17 @@ var ValidateEndpointSliceName = apimachineryvalidation.NameIsDNSSubdomain
 // ValidateEndpointSlice validates an EndpointSlice.
 func ValidateEndpointSlice(endpointSlice *discovery.EndpointSlice) field.ErrorList {
 	allErrs := apivalidation.ValidateObjectMeta(&endpointSlice.ObjectMeta, true, ValidateEndpointSliceName, field.NewPath("metadata"))
+	allErrs = append(allErrs, validateEndpoints(endpointSlice.Endpoints, *endpointSlice.AddressType, field.NewPath("endpoints"))...)
+	allErrs = append(allErrs, validatePorts(endpointSlice.Ports, field.NewPath("ports"))...)
 
-	// AddressType should have had a default value set at this point, this is
-	// just a safety check if for some reason that changes or doesn't work.
-	addrType := discovery.AddressType("")
-	if endpointSlice.AddressType == nil {
-		allErrs = append(allErrs, field.Required(field.NewPath("addressType"), ""))
-	} else {
-		addrType = *endpointSlice.AddressType
-	}
+	return allErrs
+}
 
-	if !supportedAddressTypes.Has(string(addrType)) {
-		allErrs = append(allErrs, field.NotSupported(field.NewPath("addressType"), addrType, supportedAddressTypes.List()))
-	}
+// ValidateEndpointSliceCreate validates an EndpointSlice when it is created.
+func ValidateEndpointSliceCreate(endpointSlice *discovery.EndpointSlice) field.ErrorList {
+	allErrs := ValidateEndpointSlice(endpointSlice)
 
-	allErrs = append(allErrs, validateEndpoints(endpointSlice.Endpoints, addrType, field.NewPath("endpoints"))...)
-	allErrs = append(allErrs, validatePorts(endpointSlice.Ports, field.NewPath("ports"))...)
+	allErrs = append(allErrs, validateAddressType(endpointSlice.AddressType, supportedAddressTypes)...)
```

Comment:
You have to validate this in sequence, otherwise you will validate all endpoints against an invalid address.

Looking at that code, it will fall through the switch, which is OK, but a little non-obvious.  I would instead do something like:

```
validateEndpointSlice(slice, allowedAddressTypes)
```

so teh create and update paths can pass different values.  Then just verify it in there.  For Update, check immutable after that returns (as it is).

Reply:
Given the suggestion above not to revalidate address type on update, does it still make sense to pass in `allowedAddressTypes` to `validateEndpointSlice`? Since the validation only validates known address types, an invalid address type value should not cause validation to fail as far as I can tell. Would it be better to return immediately if an address type is invalid?

Reply:
I would still make that change - I think it makes the code clearer, and when "IP" gets removed, we can remove that argument entirely.

I would also add a comment at the top of the switch along the lines of "This validates known address types, unknown types fall through and do not get validated", and at the bottom of the switch, indicating that fall-through scenario.  So no well-meaning person adds a default block.

That way, invalid address types will flag the address type, but valid ones will flag invalid values.

Reply:
I've updated the implementation to reflect what I think you meant, but let me know if I was off.